fix(entropy): Improve UX of entropy creation process (SQSERVICES-1674)

comawill opened the pull request:
<!--do not remove this marker, its needed to replace info when ticket title is updated -->
<!--jira-description-action-hidden-marker-start-->

<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/SQSERVICES-1674" title="SQSERVICES-1674" target="_blank"><img alt="Bug" src="https://wearezeta.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10803?size=medium" />SQSERVICES-1674</a>  Entropy creation UI is misleading and indicates that the App might generate "bad" entropy
  </td></table>
  <br />
 

<!--jira-description-action-hidden-marker-end-->
<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³
----

# What's new in this PR?

This PR is a draft to discuss a possible solution to improve the UX of the entropy creation process.

comawill committed:
fixup: make progressbar respond instantly

comawill committed:
fixup(entropy): don't store null value

comawill committed:
fixup(entropy): don't store duplicate cursor position values

comawill committed:
fix(entropy): Improve UX of entropy creation process

comawill committed:
fixup(entropy): new behavior of canvas affects tests

tlebon commented on src/script/auth/component/EntropyCanvas.tsx:
```diff
@@ -35,35 +35,36 @@ const EntropyCanvas = (props: CanvasProps) => {
+    if (percent >= MIN_FRAMES) {
       clearInterval();
     }
   }, [percent, percent]);
```
hmm i guess i wrote this but now I'm wondering why the dependency array has two copies of this

tlebon replied:
```suggestion
  }, [clearInterval, percent]);
```

tlebon commented on src/script/auth/page/EntropyContainer.tsx:
```diff
@@ -88,7 +88,7 @@ const EntropyContainer = ({onSetEntropy}: Props) => {
         </>
       )}
-      {percent === 95 && frames < 300 && <Muted center>{_(setEntropyStrings.moreEntropyNeeded)}</Muted>}
+      {percent >= 95 && percent < 100 && <Muted center>{_(setEntropyStrings.moreEntropyNeeded)}</Muted>}
```
i think this should just be removed, doesn't make sense to me anymore

tlebon commented on src/script/auth/component/EntropyCanvas.tsx:
```diff
@@ -35,35 +35,36 @@ const EntropyCanvas = (props: CanvasProps) => {
     onProgress(entropy, percent, true);
     pauseInterval();
-
-    setEntropy([...entropy, null]);
```
this should still be in here, creates a UI bug now if you leave the box but come back on a different side.

comawill replied:
ah... I understand, then this needs to be some how recognized when counting the frames, as this increments the size of the array without adding data

tlebon commented on src/script/auth/component/ProgressBar.tsx:
```diff
@@ -46,7 +46,6 @@ export const ProgressBar = ({width, percent, error, style}: ProgressProps) => {
           backgroundColor: error ? 'red' : 'blue',
           borderRadius: '5px',
           height: 'inherit',
-          transition: '0.5s ease-in',
```
i think this makes it appear nicer also, id probably leave this

comawill replied:
With the transition, the change is horribly delayed, and the progress bar still moves for some time after the mouse movement was stopped. This is probably the fault of the many small incremental updates. Maybe there is a better solution, so I'm definitely open to suggestions.

tlebon replied:
maybe ill play around with this a bit and see if there's some better solution

tlebon commented on src/script/auth/component/EntropyCanvas.tsx:
```diff
@@ -35,35 +35,36 @@ const EntropyCanvas = (props: CanvasProps) => {
 
   const {clearInterval, startInterval, pauseInterval} = usePausableInterval(() => {
     // This prevents automatic batching of state updates from react 18. https://github.com/reactwg/react-18/discussions/21
     flushSync(() => {
```
as far as i can tell, flushSync is no longer necessary. should be removed
```suggestion

```

tlebon commented on src/script/auth/component/EntropyCanvas.tsx:
```diff
@@ -35,35 +35,36 @@ const EntropyCanvas = (props: CanvasProps) => {
+        setLastFramesCount(frames);
+        setPercent(Math.ceil(Math.min(timedPercent, (100 * frames) / MIN_FRAMES)));
+      }
     });
```
the closing of flushSync should also be removed
```suggestion

```

comawill committed:
fixup(entropy): remove notification that more entropy is needed

comawill committed:
refactor(entropy): define class to collect entropy with and provide
uint8array data in one dedicated place

comawill committed:
Merge branch 'dev' into comawill/entropy

comawill committed:
fix(entropy): enable entropy only if explicitly enabled

add additional environment variable to enforce entropy on non windows
clients for testing

comawill committed:
fix(entropy): improve UI description texts (SQSERVICES-1675)

comawill committed:
fix(entropy): use Shannon entropy as constraint to finish entropy
creation

comawill committed:
fix(entropy): make canvas bigger

comawill committed:
fixup(entropy): improve entropy calculation by analyzing the delta
between previous values

comawill committed:
Merge branch 'dev' into comawill/entropy2

comawill committed:
fixup(entropy): include time value into EntropyFrame

tlebon commented on src/script/auth/component/EntropyCanvas.tsx:
```diff
@@ -19,89 +19,90 @@
 const EntropyCanvas = (props: CanvasProps) => {
-  const {sizeX, sizeY, onSetEntropy, onProgress, css, ...rest} = props;
-  const canvasRef = useRef(null);
+  const {sizeX, sizeY, onProgress, css, min_entropy_bits, min_frames, ...rest} = props;
```
can we change these to camelCase
```suggestion
  const {sizeX, sizeY, onProgress, css, minEntropyBits, minFrames, ...rest} = props;
```

tlebon commented on src/script/auth/component/EntropyCanvas.tsx:
```diff
@@ -19,89 +19,90 @@
-    flushSync(() => {
-      setPercent(percent => percent + 1);
-    });
+    setPercent(Math.floor(100 * Math.min(entropy.entropyBits / min_entropy_bits, entropy.length / min_frames)));
```
```suggestion
    setPercent(Math.floor(100 * Math.min(entropy.entropyBits / minEntropyBits, entropy.length / minFrames)));
```

tlebon commented on src/script/auth/page/EntropyContainer.tsx:
```diff
@@ -77,18 +77,18 @@ const EntropyContainer = ({onSetEntropy}: Props) => {
+            sizeY={512}
             onProgress={onProgress}
+            min_entropy_bits={1024}
+            min_frames={300}
```
```suggestion
            minEntropyBits={1024}
            minFrames={300}
```

tlebon commented on src/script/strings.ts:
```diff
@@ -292,17 +292,17 @@ export const setEmailStrings = defineMessages({
   subheadline: {
     defaultMessage:
-      'Move your mouse as randomly as possible within the window below until the progress bar is 100% filled. This creates a random number which helps to protect your account and increases the overall security.',
+      'Move your mouse as randomly as possible within the window below until the progress bar is 100% filled. This increases the quality of the random numbers which are used to create your cryptographic long term secrets and therefore increases the overall security of your Wire client.',
```
this text was previously cleared with content, thus new message may not be as understandable. @Astrid-Wire 
(keep in mind the en-us.json value will still appear over this)

comawill replied:
A proposal to change the translated text (en and de) was made on crowdin as well

tlebon commented on src/script/auth/component/EntropyCanvas.tsx:
```diff
@@ -19,89 +19,90 @@
-    setEntropy(entropy => [...entropy, newEntropy]);
+    const boundingRect = event.currentTarget?.getBoundingClientRect();
+    const drawPoint: EntropyFrame = {
+      t: 0,
```
is there a reason for t to be included in here besides its inclusion in the type? i suggest to make it an optional parameter in the type or use a different type perhaps.

comawill replied:
it was mainly because of the type, made it optional now

tlebon commented on src/script/util/Entropy.test.ts:
```diff
@@ -0,0 +1,98 @@
+
+import {EntropyData, shannonEntropy, calculateDeltaValues} from './Entropy';
+
+//import {joaatHash} from './Crypto';
```
maybe remove comment

tlebon commented on src/script/util/Entropy.ts:
```diff
@@ -0,0 +1,97 @@
+}
+
+export class EntropyData {
+  readonly frames: EntropyFrame[];
```
are we assuming that this will always be our entropy type? perhaps this may also have a generic option?

comawill replied:
At the moment yes. EntropyData and EntropyFrame are directly coupled especially for the assumption how to calculate the estimated entropy, so making this generic would require a different data model

comawill committed:
fixup(entropy): fix naming and remove debugging leftovers

Co-authored-by: Timothy LeBon <tlebon@gmail.com>

tlebon commented on src/i18n/en-US.json:
```diff
@@ -1081,9 +1081,10 @@
   "setEntropy.headline": "Create Entropy",
   "setEntropy.moreEntropyNeeded": "Please continue moving your mouse within the window to create enough entropy.",
-  "setEntropy.subheadline": "Move your mouse as randomly as possible within the window below until the progress bar is 100% filled. This creates a random number which helps to protect your account and increases the overall security.",
+  "setEntropy.subheadline": "Move your mouse as randomly as possible within the window below until the progress bar is 100% filled. This increases the quality of the random numbers which are used to create your cryptographic long term secrets and therefore increases the overall security of your Wire client.",
```
I'd still like @Astrid-Wire s comments on these

Astrid-Wire replied:
```suggestion
  "setEntropy.subheadline": "Move your mouse as randomly as possible within the window below until the progress bar is 100% filled. This increases the quality of the random numbers which are used to create your cryptographic long term secrets and therefore increases the overall security of this device.",
```

tlebon commented on src/script/Config.ts:
```diff
@@ -43,7 +43,8 @@ export class Configuration {
     ...env.FEATURE,
-    ENABLE_EXTRA_CLIENT_ENTROPY: env.FEATURE.ENABLE_EXTRA_CLIENT_ENTROPY && Runtime.isWindows(),
+    ENABLE_EXTRA_CLIENT_ENTROPY:
+      env.FEATURE.ENABLE_EXTRA_CLIENT_ENTROPY && (Runtime.isWindows() || env.FEATURE.ENABLE_EXTRA_CLIENT_ENTROPY_FORCE),
```
Can we rename this? Enable and force together make less sense to me. 
```suggestion
      env.FEATURE.ENABLE_EXTRA_CLIENT_ENTROPY && (Runtime.isWindows() || env.FEATURE.FORCE_EXTRA_CLIENT_ENTROPY),
```

tlebon commented on src/script/auth/component/EntropyCanvas.tsx:
```diff
@@ -19,89 +19,89 @@
 const EntropyCanvas = (props: CanvasProps) => {
-  const {sizeX, sizeY, onSetEntropy, onProgress, css, ...rest} = props;
-  const canvasRef = useRef(null);
+  const {sizeX, sizeY, onProgress, css, minEntropyBits, minFrames, ...rest} = props;
```
I'm wondering if it makes sense to include a suggested/default value or JsDoc for any laymen who may want to use this in the future but has less understanding of bits or Frames

tlebon replied:
This may be covered in your comment on the interface

tlebon reviewed (approved):
Lgtm

comawill committed:
fixup(entropy): rename ENABLE_EXTRA_CLIENT_ENTROPY_FORCE to FORCE_EXTRA_CLIENT_ENTROPY

comawill committed:
fixup(entropy): improve wording in setEntropy.subheadline

Co-authored-by: Astrid Pahl <36076182+Astrid-Wire@users.noreply.github.com>

comawill committed:
fixup(entorpy): update jsdoc
